Python3 fork tidy copy

wxtim opened the pull request:
PR comparing my GUI removed Python 2 Rose   with Python3 Rose :snake: :snake: :snake: :rose:

ThomasColemanBoM committed:
Extend rose_bunch to update bunch-arg values

By using itertools, we can update the values for different bunch-args
to shrink, extend, or create new permutations. The ones used so far
include 'izip', 'izip_longest', and 'product'.

wxtim committed:
initial removal of rose and rosie gui

oliver-sanders committed:
remove docs latex dependency

oliver-sanders committed:
cylc_lang: support inter-suite-triggers

oliver-sanders committed:
cylc_lang: support include files

oliver-sanders committed:
cylc_lang: support xtriggers

oliver-sanders committed:
cylc_lang: support empy + simplify jinja2

oliver-sanders committed:
cylc_lang: support line continuation in headings

Commit:
Fix issue by adjusting the logic

oliver-sanders committed:
use python implementation of which

Commit:
Remove pointless message

paulcresswell committed:
Merge pull request #2275 from stevewardle/rose_ana_kgo_info_fix

Fix overzealous informational message about KGO in grepper class

oliver-sanders committed:
cylc_lang: support suicide triggers

oliver-sanders committed:
rosie: fix traceback from importing gi.repository.Secret

oliver-sanders committed:
fix rose suite-shutdown

sadielbartholomew committed:
Merge pull request #2273 from oliver-sanders/make-latex-history

remove docs latex dependency

sadielbartholomew committed:
Merge pull request #2277 from oliver-sanders/rose-suite-shutdown

rose suite-shutdown: fix traceback

sadielbartholomew committed:
Merge pull request #2276 from oliver-sanders/rosie-secret-import-error

rosie: fix traceback from importing gi.repository.Secret

wxtim committed:
isodatetime temporarily removed

wxtim committed:
Manual conversion of
* rose config
* rose app-run
* rose config-diff
* rose config-dump
* rose env-cat
* rose namelist-dump
* rose metadata-check (questionable test results)
* rose metadata-gen
* rose metadata-graph
* rose host-select

Automated 2to3 changes

Automatic replacement of references to python2 with python3
Exceptions are:
* CHANGES.md
* bin/rose-make-docs
* sphinx/installation.rst

@2to3 -f import

@2to3 -f basestring

@2to3 -w -x dict -x filter -x map -x idiom -x zip -x tuple_params -x xrange

rose config made to work

rose-app-run made to work

rose config-diff, env-cat and namelist-dump working

rose host-select and metadata-* checked and working

rose-config-dump made to work

oliver-sanders committed:
cylc_lang: support cycle points in inter-cycle offsets

wxtim commented on bin/rose-check-software:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2
+#!/usr/bin/env python3
```
#fixme : Whilst the shebang line has been changed this script is still :snake: 2 and :. not ok.

wxtim commented on etc/tutorial/cylc-forecasting-suite/lib/python/util.py:
```diff
@@ -178,7 +178,7 @@ def plot_vector_grid(filename, x_grid, y_grid):
         matplotlib.use('Agg')
```
Also fixed a spelling error because I'm pedantic like that.

wxtim commented on lib/python/rose/app_run.py:
```diff
@@ -472,7 +472,7 @@ def main():
         runner.handle_event(exc)
         if opts.debug_mode:
-            traceback.print_exc(exc)
+            traceback.print_exc()
```
@mshin tells me that this was always wrong!

wxtim commented on lib/python/rose/job_runner.py:
```diff
@@ -183,36 +183,53 @@ def run(self, job_manager, *args):
-                results[job.name] = result
-            if results:
-                sleep(self.POLL_DELAY)
+#         nproc = self.nproc
```
I had real problems getting multiprocessing.pool to work: @mattshin had to disable it, at least for now, and this section will need to be re-written in  a python3 friendly way.

The problem seems to be caused by python 3 being unable to serialize some objects and therefore to pass them to processes. :(

mattshin replied:
Hey! Just wanted to lyk that you probably meant to tag someone else in this comment 😅 - maybe @mshin?

wxtim replied:
Yes!

wxtim replied:
Thank you @mattshin, I did mean @mshin

wxtim commented on sbin/rosa-rpmbuild:
```diff
@@ -67,7 +67,7 @@ License: GPLv3
 URL: https://github.com/metomi/$NAME/
```
This probably needs some work....

wxtim commented on sphinx/Makefile:
```diff
@@ -60,7 +60,7 @@ slides:
 	@echo "Build finished. The slide-html files are in $(BUILDDIR)/slides."
```
Sphinx stuff need real thought...

oliver-sanders replied:
I wouldn't worry about the docs for now.

wxtim committed:
small corrections

oliver-sanders commented on lib/python/rose/apps/rose_ana_v1.py:
```diff
@@ -464,7 +465,7 @@ def load_user_comparison_modules(self, files):
             for obj_name, obj in contents:
                 att_name = "run"
-                if hasattr(obj, att_name) and callable(getattr(obj, att_name)):
+                if hasattr(obj, att_name) and isinstance(getattr(obj, att_name), collections.Callable):
```
Looks like `callable` dissipated at 3.0 and [returned at 3.2](https://docs.python.org/3/library/functions.html#callable) so we can probably revert this diff.

oliver-sanders commented on lib/python/rose/bush.py:
```diff
@@ -34,7 +34,7 @@
 from time import gmtime, strftime
 import traceback
-import urllib
+import urllib.request, urllib.parse, urllib.error
```
Line contains unused imports?

oliver-sanders commented on lib/python/rose/config.py:
```diff
@@ -1062,15 +1062,16 @@ def dump(self, root, target=sys.stdout, sort_sections=None,
-        root_keys = root.value.keys()
-        root_keys.sort(sort_sections)
+        root_keys = list(root.value.keys())
+        root_keys.sort(key=cmp_to_key(sort_sections))
```
Worth sticking a TODO on these lines to remind us to address the `cmp_to_key` issue in the near future.

wxtim replied:
Good Idea - although my thoughts were to put it near the import statement - What do you think?

oliver-sanders replied:
Either or.

oliver-sanders commented on lib/python/rose/config.py:
```diff
@@ -1377,7 +1388,12 @@ def load(self, source, node=None, default_comments=None):
+                pos = handle.tell()
+                line = handle.readline().decode()
+            except:
+                handle.seek(pos)
```
Does this not auto-rewind the file?

wxtim replied:
Apparently not, although I've changed it in the latest version because there were other issues...

oliver-sanders commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -245,8 +245,9 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
-                    for prev_path, path in zip(
-                            prev_target.paths, target.paths):
+                    prev_target.paths.sort()
+                    for prev_path, path in list(zip(
```
Is this `list` necessary? My understanding is that we only need to go down this route if we modify the iterable within the loop.

wxtim replied:
I did not let 2to3 do this - I must have done thsi manually, though I'm not sure why...

oliver-sanders replied:
If we modify the thing we are iterating over nasty things happen:

```python
>>> dic = {'a': 1}
>>> for item in dic:
...     dic.pop(item)
... 
1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dictionary changed size during iteration
```

No modification here so I think we're safe.

wxtim committed:
Re-added isodatetime from python3 version

changes to make rose-app-upgrade work

added ModuleNotFoundError to tuple with ImportError in all cases

changes to make rose-stem tests pass

dictionary changed size error fixed

small change to popen

done tests for rose.config_tree

made rose.variable work - caused by dict ordering issues

rose-suite-cmp-vc now working

rose-macro: test 00 fixed

rose-macro: fixed test 01

rose-macro: fixed test 02

oliver-sanders committed:
report debug statements only when requested

sadielbartholomew committed:
Update t/rose-task-run/31-app-bunch.t

ARGS for argument_mode_izip_longest were incorrect

Co-Authored-By: ColemanTom <igenous@gmail.com>

ThomasColemanBoM committed:
31-app-bunch.t pull patterns to top of file to shorten lines

ThomasColemanBoM committed:
Provide more explanation about the tests approach in 31-app-bunch.t for the itertools tests

ThomasColemanBoM committed:
Add link to itertools documentation in rose_bunch.rst

oliver-sanders committed:
docs: patch hieroglyph to work with sphinx 1.8.0+

sadielbartholomew committed:
Merge pull request #2279 from oliver-sanders/debug-mode

report debug statements only when requested

sadielbartholomew committed:
Merge pull request #2280 from oliver-sanders/hieroglyph-hack

docs: patch hieroglyph to work with sphinx 1.8.0+

ThomasColemanBoM committed:
Fix ARG1 izip test in 31-app-bunch

sadielbartholomew committed:
Merge pull request #2265 from ColemanTom/rose-bunch_all_permutations_option

Extend rose_bunch with itertools

wxtim committed:
Rose Macro Now working
Rose bush and its test removed

wxtim committed:
rose-app-upgrade fixed

wxtim commented on etc/tutorial/cylc-forecasting-suite/bin/consolidate-observations:
```diff
@@ -30,7 +30,7 @@ def plot_wind_data(wind_x, wind_y, x_range, y_range, x_coords, y_coords,
         matplotlib.use('Agg')
         import matplotlib.pyplot as plt
-    except ImportError:
+    except (ImportError, ModuleNotFoundError):
```
Not entirely convinced that the `ImportError` is still required.

wxtim commented on lib/python/rose/cmp_source_vc.py:
```diff
@@ -89,7 +90,7 @@ def main():
     try:
         lines = suite_vc_cmp.cmp_source_vc_info(suite_name=suite_name)
-    except (StandardError, RosePopenError) as exc:
+    except (Exception, RosePopenError) as exc:
```
This feels bad - Should this be more specific?

wxtim commented on lib/python/rose/config.py:
```diff
@@ -97,6 +97,7 @@
 import shlex
 import sys
 from tempfile import NamedTemporaryFile, TemporaryFile
+from functools import cmp_to_key
```
The consequences of this in terms of performance are as yet unclear.

wxtim commented on lib/python/rose/config.py:
```diff
@@ -1079,7 +1080,7 @@ def dump(self, root, target=sys.stdout, sort_sections=None,
                 section_keys.append(key)
         if root_option_keys:
-            handle.write(blank)
+            self._write_safely(handle, blank)
```
Is it worth making this a module method, or just leaving it as a `@staticmethod` of this class?

wxtim commented on lib/python/rose/config.py:
```diff
@@ -1118,26 +1120,39 @@ def _string_node_dump(self, key, node, handle, env_escape_ok):
         """Return text representation of a configuration comment."""
         return "#%s\n" % (comment)
 
+    @staticmethod
```
I suggest that I replace this with the code in my gist at https://gist.github.com/wxtim/cc3f4e653d37616ff24714bfbc32444b.

wxtim commented on lib/python/rose/config.py:
```diff
@@ -1324,16 +1339,17 @@ def load_defines(self, defines, node=None):
         # "ignore" flags. Rather than replicating the logic for parsing
         # ignore flags, it is actually easier to write the values in
         # "defines" to a file and pass it to the loader to parse it.
+        #source = TemporaryFile(mode='w', encoding='UTF-8')
```
@ TODO - remove?

wxtim commented on lib/python/rose/suite_engine_procs/cylc.py:
```diff
@@ -75,6 +75,14 @@ def check_global_conf_compat(self):
                     "[hosts][localhost]work directory"]:
             out = self.popen("cylc", "get-global-config", "-i", key)[0]
             lines = out.splitlines()
+            try:
```
This logic is odd, and should probably use if/else logic.

wxtim committed:
put sphinx back to master for now

oliver-sanders committed:
fix test: ensure cylc-run exists before running rose tutorial cmds

matthewrmshin committed:
Merge pull request #2282 from oliver-sanders/flaky-test-docs-02

fix flaky test: docs/02

matthewrmshin committed:
Update for release + happy 2019

sadielbartholomew committed:
Merge pull request #2274 from oliver-sanders/pygments-lexer-imprevements

pygments syntax: improve cylc lexer

oliver-sanders committed:
docs: fix suite design-guide url

matthewrmshin committed:
Merge pull request #2285 from oliver-sanders/fix-suite-design-guide-url

docs: fix suite design-guide url

oliver-sanders committed:
Merge pull request #2284 from matthewrmshin/release

Update for release + happy 2019

wxtim committed:
work towards rose

wxtim committed:
corrections to isodatetime

wxtim committed:
further isodatetime changes

wxtim committed:
fix to make all the rosie-id tests work

wxtim committed:
rosie delete fixed

wxtim committed:
rosie-create tests 1 & 3 fixed

wxtim committed:
rosie create also working

wxtim committed:
some work on rosie ls.

wxtim committed:
fix rosa-svn-pre-commit

matthewrmshin committed:
config-edit: fix int spin box manual edit

Value change is not being picked up when users input text to the int spin box.
This change should fix.

matthewrmshin committed:
Call opendir before testing contact file existence

Call `os.open(..., os.O_DIRECTORY)` on the suite's service directory
(i.e. the parent directory of the suite's contact file) before testing
for existence of the contact file. This ensures that the file system is
synchronised e.g. for NFS or the file existence test may fail even
though the file is already there.

oliver-sanders committed:
Merge pull request #2290 from matthewrmshin/opendir-before-get-contact-file

Call opendir before testing contact file existence

wxtim committed:
fixed some stuff

sadielbartholomew committed:
Merge pull request #2289 from matthewrmshin/fix-config-edit-int-spin-box-edit

config-edit: fix int spin box manual edit

wxtim committed:
changed travis to 3.6

wxtim committed:
cmp work

wxtim committed:
changes to remove cmp statements

oliver-sanders commented:
@wxtim you might want to close this PR now that it has been raised on metomi/rose

wxtim committed:
changes to address issues OS brought up

wxtim committed:
Oliver's suggested changes to rose-check-software

wxtim committed:
Changes to travis.yml

wxtim committed:
travis yml changes
